src/components/HighlightText.js

```diff
@@ -0,0 +1,153 @@
+import { isNil } from 'lodash';
+import React, {useCallback} from 'react';
+
+/**
+ * Checks within a text that the characters between
+ * startIdx and endIdx are complete words.
+ *
+ * E.g.,
+ *    text="Biometric", startIdx=3 (m), endIdx=8 (c) => false
+ *    text="The metric", startIdx=4 (m), endIdx=9 (c) => true
+ */
+const punctuation = ['.', ',', ':', ';', '!', '?'];
+const isStandaloneWord = (text, startIdx, endIdx) => {
+  if (startIdx !== 0) {
+    const char = text.charAt(startIdx-1);
+
+    if (char.trim() !== '' && !punctuation.includes(char)) {
+      return false;
+    }
+  }
+
+  if (endIdx !== text.length-1) {
+    const char = text.charAt(endIdx);
+
+    if (char.trim() !== '' && !punctuation.includes(char)) {
+      return false;
+    }
+  }
+
+  return true;
+};
+
+
+const findHighlightableMatch = (highlight, text, idx) => {
+  for (const highlightConfig of highlight) {
+    for (const word of highlightConfig.words) {
+      const textAtCurrWordLength = text.slice(idx, idx+word.length);
+
+
+      const textIsStandalone = isStandaloneWord(text, idx, idx+word.length);
+      const textMatchesHighlightableWord = textAtCurrWordLength.toLowerCase() === word.toLowerCase();
+
+      if (textIsStandalone && textMatchesHighlightableWord) {
+        return {
+          word: textAtCurrWordLength,
+          bgColor: highlightConfig.bgColor,
+          textColor: highlightConfig.textColor,
+        };
+      }
+    }
+  }
+
+  return null;
+};
+
+export const HighlightText = (props) => {
+  const {
+    /**
+     * highlight:
+     *   Specifies any number of highlight colors and words
+     *   to highlight with those colors.
+     * schema:
+     *   [
+     *    { bgColor: ..., textColor: ... , words: [ ... ] },
+     *    { bgColor: ..., textColor: ... , words: [ ... ] },
+     *   ]
+     */
+    highlight,
+    text,
+  } = props;
+
+  // searches through text one character at a time
+  // looking for words to highlight
+  const splitAndHighlight = useCallback(() => {
+    if (isNil(text)) {
+      return <div></div>;
+    }
+
+    const output = [];
+
+    const pushOutput = ({
+      text,
+      bgColor=undefined,
+      textColor=undefined,
+      fontWeight=undefined
+    }) => {
+      return (
+        <span
+          key={output.length}
+          data-cy={`highlight-${output.length}`}
+          style={{
+            backgroundColor: bgColor,
+            color: textColor,
+            fontWeight: fontWeight
+          }}>
+          {text}
+        </span>
+      );
+    };
+
+
+    // keeps track of last unhighlighted text
+    let endOfLastHighlight = 0;
+
+    // scan through text by character
+    for (let idx = 0; idx < text.length; idx++) {
```

Comment:
Why not search with regular expressions here?

I spent a good time thinking about this one.  The best example around what I was thinking about that I could find was in this repo here: https://github.com/bvaughn/highlight-words-core/blob/master/src/utils.js

Reply:
I assumed it would need each word to be unique / not a subword. For example “Origin” is in the word list, and we wouldn’t want to highlight origin in “original”. I assumed, if we wanted to include plurals such as diseases, we could just add it to the word list, but maybe we should check in with Jonathan first. The big issue with regex for me was that a lot of browsers do not support positive lookbehind, which is the only real way to ensure that a word is surrounded by whitespace to ensure that it is its own word. It’s still possible to use regex, but we would have to manually filter the results if we wanted to make sure that sub-words are not highlighted, which seemed clunky to me. However, if we wanted to highlight subwords, then regex is much easier.

Reply:
> lot of browsers do not support positive lookbehind

Huh - I had to look this up: https://www.lambdatest.com/web-technologies/js-regexp-lookbehind
Looks like this is true. For the cases we really care about (Chrome, newest FF), it is fully supported. On a policy note, we [officially support Chrome](https://support.terra.bio/hc/en-us/articles/360028235911-How-to-register-for-a-Terra-account). I'm fine with the char-by-char approach but would love to see some performance metrics for long RUS statements.

Reply:
Following up on my question, I used [this example RUS](https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/GetPdf.cgi?id=phd001030) in DAR-601 (dev). The timing for a rather large RUS seems pretty reasonable. Highlighting takes the most time in rendering that element, which makes sense.
 
<img width="1791" alt="Screen Shot 2023-01-09 at 7 34 31 AM" src="https://user-images.githubusercontent.com/116679/211309800-524b556f-25a8-499e-bc35-c573ebfb071b.png">

Reply:
I am more than happy to look into making it more performant, maybe using some hashmaps or mixing regex and manual search. But if it's good enough as is, then I am happy with merging too.

Reply:
> But if it's good enough as is, then I am happy with merging too.

I do, thank you!